### PR TITLE
Update boringssl and boringssl-with-bazel to latest master

### PR DESCRIPTION
**Pull request by jtattermusch**
supersedes https://github.com/grpc/grpc/pull/18751/files

Fixes https://github.com/grpc/grpc/issues/20037

### CODE REVIEW

**Comment by jtattermusch** [comment]
@muxi  I'm not sure why, but the new version of boringssl seems to be generating the symbols in a slightly different format, so running `tools/distrib/generate_grpc_shadow_boringssl_symbol_list.sh` ends up in an empty list of symbols to shadow.
I found that the symbols included in the previous version are no longer prefixed by `_`
https://github.com/grpc/grpc/blob/21cf17aacf79e635fe81a0a629d2588eda0441cc/tools/distrib/generate_grpc_shadow_boringssl_symbol_list.sh#L36

What is the right resolution here?

**Comment by muxi** [comment]
Agree that this is fragile, and if BoringSSL team started to do different things it might break us. I think the right solution is probably start to use BoringSSL's [new feature](https://boringssl.googlesource.com/boringssl/+/refs/heads/master/BUILDING.md) `BORINGSSL_PREFIX_SYMBOLS` to add the prefix rather than gRPC

**Comment by jtattermusch** [comment]
quite a few tests broken by this PR unfortunately

**Comment by jtattermusch** [comment]
The problem is that newer versions of boringssl now use the  $<COMPILE_LANGUAGE:CXX> construct, which doesn't work with Visual Studio Generators.
We'll need to fix that and backport the fix to chromium-stable branch first.

```
--     T:/src/github/grpc/workspace_csharp_windows_dbg_native_default_coreclr/third_party/zlib/zconf.h
-- to 'zconf.h.included' because this file is included with zlib
-- but CMake generates it automatically in the build directory.
-- Configuring done
CMake Error at third_party/boringssl/CMakeLists.txt:244 (add_compile_options):
  Error evaluating generator expression:
    $<COMPILE_LANGUAGE:CXX>
  $<COMPILE_LANGUAGE:...> may not be used with Visual Studio generators.
CMake Error at third_party/boringssl/CMakeLists.txt:244 (add_compile_options):
  Error evaluating generator expression:
    $<COMPILE_LANGUAGE:CXX>
  $<COMPILE_LANGUAGE:...> may not be used with Visual Studio generators.
CMake Error at third_party/boringssl/CMakeLists.txt:244 (add_compile_options):
  Error evaluating generator expression:
    $<COMPILE_LANGUAGE:CXX>
  $<COMPILE_LANGUAGE:...> may not be used with Visual Studio generators.
```
https://github.com/google/boringssl/blob/2e0d354690064c90ee245c715b92e2bb32492571/CMakeLists.txt#L244

**Comment by jtattermusch** [comment]
What still doesn't work: 
- python artifact build on linux: https://source.cloud.google.com/results/invocations/bbd66834-cce2-48dc-9ef6-464f32b93319/targets/github%2Fgrpc/tests
- C# artifact on windows (the assembly enabled compilation with ninja seems broken): https://source.cloud.google.com/results/invocations/933b777c-5918-45a5-8c21-1a13ab22b480/targets/github%2Fgrpc/tests
- boringssl symbol shadowing is broken

**Comment by jtattermusch** [comment]
@davidben 
can you please provide some insights on the breakages:

1. the assembly enabled compilation with ninja on windows seems broken: https://source.cloud.google.com/results/invocations/933b777c-5918-45a5-8c21-1a13ab22b480/targets/github%2Fgrpc/tests
```
T:\src\github\grpc\workspace_csharp_ext_windows_x64\cmake\build\x64\third_party\boringssl\crypto\fipsmodule\p256-x86_64-asm.asm:800: error: instruction expected after label
T:\src\github\grpc\workspace_csharp_ext_windows_x64\cmake\build\x64\third_party\boringssl\crypto\fipsmodule\p256-x86_64-asm.asm:801: error: instruction expected after label
...
```

2. python artifact build on linux seems broken

https://source.cloud.google.com/results/invocations/bbd66834-cce2-48dc-9ef6-464f32b93319/targets/github%2Fgrpc/tests
```
In file included from third_party/boringssl/crypto/hrss/hrss.c:28:0:
/opt/rh/devtoolset-2/root/usr/lib/gcc/i386-CentOS-linux/4.8.2/include/emmintrin.h:31:3: error: #error "SSE2 instruction set not enabled"
 # error "SSE2 instruction set not enabled"
   ^
```

3. the naming of public boringssl symbols have changed which is causing trouble with the symbol shadowing approach we are using for gRPC objC. (see comment above:  https://github.com/grpc/grpc/pull/20100#issuecomment-525919900)

**Comment by agl** [comment]
(Echoing an internal email:)

> the assembly enabled compilation with ninja on windows seems broken: 

I think I recall correctly that yasm is abandonware now and that nasm is the correct assembler for Windows.

> python artifact build on linux seems broken

SSE2 came with the Pentium 4 but I think the code should still compile without it. The fact that including the header causes an error seems like a compiler bug.

You could pass -msse2 to enable support. The issue in the GCC headers seems to have been fixed with [this thread](https://gcc.gnu.org/ml/gcc-patches/2013-04/msg00740.html), which was released in 4.9.0. We follow the [absl policy](https://abseil.io/about/compatibility#what-we-promise) here and maintain hacks for five years, but 4.9 was released more than five years ago, so we would not typically make a change for this.

**Comment by davidben** [comment]
> I think I recall correctly that yasm is abandonware now and that nasm is the correct assembler for Windows.

Comparing the line numbers against the file, it seems to all be `adcx` and `adox` instructions. That suggests your yasm is too old to understand ADX instructions.
https://boringssl.googlesource.com/boringssl/+/a21f78d24bf645ccd6774b2c7e52e3c0514f7f29/win-x86_64/crypto/fipsmodule/p256-x86_64-asm.asm#800

Given you need to update the assembler on those bots anyway, just switch to nasm. yasm hasn't made a release for five years now, so even the latest one is missing instructions. Once we start using them, yasm will stop working.

**Comment by jtattermusch** [comment]
FTR, this is where boringssl has switched to recommending NASM instead of YASM.
https://github.com/google/boringssl/commit/73d69f4d51086da93ea10c998b14e05611910972 (we'd need to do the same).

**Comment by jtattermusch** [comment]
@agl  @davidben  thanks for help with diagnosing the yasm issue - I'll try to switch our build to use NASM. 

I'll also see if I can do something about the pre-gcc4.9 problem (the python artifacts).

Do you have any insights on the symbol naming?  (problem 3. here: https://github.com/grpc/grpc/pull/20100#issuecomment-529392044)

**Comment by davidben** [comment]
The symbol thing I'm not sure about. The C ABI is still the C ABI and not a function of BoringSSL or anything.

I would suggest printing out what nm outputs and poking at your script from there.

**Comment by jtattermusch** [comment]
I was able to switch yasm -> nasm for the windows artifact build, but ran into another problem.
It seems that the cmake `add_definition()` statements we are using in a few places collide with nasm (cmake appends them to the nasm.exe command line and that fails because nasm thinks they are input files to process).
```
C:\PROGRA~1\NASM\nasm.exe  -I..\..\..\third_party\boringssl\third_party\googletest\include -I..\..\..\third_party\boringssl\crypto\..\include -I..\..\..\third_party\boringssl\crypto\fipsmodule\..\..\include /wd4065 /wd4506 /wd4200 /wd4291 /wd4244 /wd4267 /wd4987 /wd4774 /wd4819 /wd4996 /wd4619 -f win64 -o third_party\boringssl\crypto\fipsmodule\CMakeFiles\fipsmodule.dir\aesni-x86_64.asm.obj T:\src\github\grpc\workspace_csharp_ext_windows_x64\cmake\build\x64\third_party\boringssl\crypto\fipsmodule\aesni-x86_64.asm

/wd4065: fatal: more than one input file specified: /wd4506
```
https://github.com/grpc/grpc/blob/88cffa00827e5e0299be7c51803896e49281ef5f/CMakeLists.txt#L103

**Comment by jtattermusch** [comment]
> (Echoing an internal email:)
> 
> > the assembly enabled compilation with ninja on windows seems broken:
> 
> I think I recall correctly that yasm is abandonware now and that nasm is the correct assembler for Windows.
> 
> > python artifact build on linux seems broken
> 
> SSE2 came with the Pentium 4 but I think the code should still compile without it. The fact that including the header causes an error seems like a compiler bug.
> 
> You could pass -msse2 to enable support. The issue in the GCC headers seems to have been fixed with [this thread](https://gcc.gnu.org/ml/gcc-patches/2013-04/msg00740.html), which was released in 4.9.0. We follow the [absl policy](https://abseil.io/about/compatibility#what-we-promise) here and maintain hacks for five years, but 4.9 was released more than five years ago, so we would not typically make a change for this.

@lidizheng   looks like the problem is that the gcc version available quay.io/pypa/manylinux1_x86_64 (the images used to build our python linux artifacts) is too old. Is there any way we can update the image?

**Comment by jtattermusch** [comment]
@muxi the problem with regenerating boringssl symbol shadow list was a false alarm - it turns out I was running the `generate_grpc_shadow_boringssl_symbol_list.sh` script on linux by mistake, which is what caused the problem, because the symbol format is different on linux. Added a check to enforce it runs on MacOS.

I still need to figure out how to fix the C# windows and python linux artifact build before I can merge.

**Comment by davidben** [comment]
> I was able to switch yasm -> nasm for the windows artifact build, but ran into another problem.
It seems that the cmake add_definition() statements we are using in a few places collide with nasm (cmake appends them to the nasm.exe command line and that fails because nasm thinks they are input files to process).

This was happening in your yasm build too. See "yasm: warning: can open only one input file, only the last file will be processed"
https://source.cloud.google.com/results/invocations/933b777c-5918-45a5-8c21-1a13ab22b480/targets/github%2Fgrpc/tests

I think the problem is your `add_definitions` lines are adding it to all languages, which does not make sense for assembly. (Those also aren't preprocessor definitions.) We add our warning lines to `CMAKE_C_FLAGS` and `CMAKE_CXX_FLAGS`.

**Comment by davidben** [comment]
Incidentally, `$<COMPILE_LANGUAGE:blah>` would also be an option here, but for the MSVC issue you ran into above. If you used an up-to-date CMake, that would also have worked fine.
https://gitlab.kitware.com/cmake/cmake/issues/17435

(An up-to-date CMake is also required to get MSVC + assembly working, #16376, since our `$<COMPILE_LANGUAGE:CXX>` had to be reverted. So I would also recommend updating that at some point. But it shouldn't block this.)

**Comment by muxi** [comment]
@jtattermusch - ok thanks for the update

**Comment by lidizheng** [comment]
@jtattermusch `manylinux1` is an image released by PyPA based on CentOS 5. If we want to be `manylinux1` compatible (make binary wheels upload-able), we need to keep it as it is.

In #20282, that is an attempt to upgrade to `manylinux2010`. But it should be considered a breaking change. I hope we can announce it and give our users sometime to adopt this change, before we completely remove `manylinux1` wheels.

**Comment by gnossen** [comment]
@lidizheng is right. We need to have a deprecation period before we can remove manylinux1 support. However, compiler version isn't necessarily equivalent to manylinux1 compatiblity. So long as we're compatible with [the manylinux1 spec](https://www.python.org/dev/peps/pep-0513/), we're good. The manylinux1 docker image hosted on `quay.io` hosts a toolchain that guarantees you compatibility. If we can find some other toolchain that works, that would be fine too.

In https://github.com/grpc/grpc/pull/20080, @veblush added a test verifying manylinux1 compatibility and in https://github.com/grpc/grpc/pull/20142, he attempted to build manylinux1 images with a different toolchain. As far as I know, that attempt was unsuccessful, but if we put enough elbow grease into it, we might be able to cobble together a toolchain with a more modern compiler.

**Comment by davidben** [comment]
Note CentOS 5 has been end-of-life for 2.5 years now (March 2017).

**Comment by njsmith** [comment]
Yeah, manylinux2010 wheels ought to work on every platform that's anywhere *near* in-support and was supported by manylinux1.

The one caveat is that to download manylinux2010 wheels, users need to be on pip 19.0 or later (released January 2019). Upgrading pip is usually pretty trivial (it's a python package, so if you can install python packages, you can upgrade pip!), but sometimes people need a nudge to do it.

Here's the stats on pip versions used this month to download your latest manylinux wheels: https://docs.google.com/spreadsheets/d/1xXrHPUnwvGMgJACMrF_P8wHFCIRrXxfhaJ7keI7d3dc/edit?usp=sharing

<details>

Query:

```
SELECT
 COUNT(*) AS downloads, REGEXP_EXTRACT(details.installer.version, "([0-9]+)\\.") AS pip_major_version
FROM
 TABLE_DATE_RANGE( [the-psf:pypi.downloads], TIMESTAMP("20190901"), CURRENT_TIMESTAMP() )
WHERE
 details.installer.name = "pip" and file.filename LIKE "grpcio-1.23.0-%manylinux1%"
GROUP BY
  pip_major_version
ORDER BY
 downloads DESC
LIMIT
 1000
```

</details>

**Comment by jtattermusch** [comment]
Another objC build problem:
```
/Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/objective-c/tests/Pods/Headers/Private/BoringSSL-GRPC/openssl_grpc/ssl.h:142:2: error: unterminated conditional directive
#ifndef OPENSSL_HEADER_SSL_H
```
https://source.cloud.google.com/results/invocations/d744a5b5-9c20-40f7-9a40-e20db70bc336/targets/github%2Fgrpc%2Frun_tests%2Fobjc_macos_opt_native%2Fios-cpp-test-cronet/tests

**Comment by jtattermusch** [comment]
Also found that the python linux artifact only fails for x86 artifacts (the 64bit ones are passing)
https://source.cloud.google.com/results/invocations/6f575475-d634-44f2-96f0-ae4aecadcf39/targets/github%2Fgrpc/tests

**Comment by njsmith** [comment]
> Also found that the python linux artifact only fails for x86 artifacts (the 64bit ones are passing)

It looks like only 0.6% of your Linux installs this month were for x86-32: https://docs.google.com/spreadsheets/d/1MbY9wguoCa5_ArntugSsw5tRAtFQohfpzhkFkwHuk6Q/edit#gid=1546023097

Also, x86-32 support is effectively deprecated in manylinux, since RH dropped support so it's impossible to maintain a build image. There's still a manylinux1 image, but it probably won't get any more updates, and there aren't any plans to provide x86-32 images for future manylinux versions.

<details>

```
SELECT
 COUNT(*) AS downloads, file.filename
FROM
 TABLE_DATE_RANGE( [the-psf:pypi.downloads], TIMESTAMP("20190901"), CURRENT_TIMESTAMP() )
WHERE
 details.installer.name = "pip" and file.filename LIKE "grpcio-1.23.0-%manylinux1%"
GROUP BY
  file.filename
ORDER BY
 downloads DESC
LIMIT
 1000
```

</details>

**Comment by jtattermusch** [comment]
Got over the cmake build problem (by removing add_definitions() for msvc warnings) but now running into another problem with Windows C# artifact build:
```
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_dll --intdir=CMakeFiles\grpc_csharp_ext.dir --manifests  -- C:\PROGRA~2\MICROS~1.0\VC\bin\amd64\link.exe /nologo CMakeFiles\grpc_csharp_ext.dir\src\csharp\ext\grpc_csharp_ext.c.obj  /out:grpc_csharp_ext.dll /implib:grpc_csharp_ext.lib /pdb:grpc_csharp_ext.pdb /dll /version:0.0 /machine:x64 /debug /INCREMENTAL  grpc.lib gpr.lib third_party\boringssl\ssl\ssl.lib third_party\boringssl\crypto\crypto.lib third_party\zlib\zlibstatic.lib third_party\cares\cares\lib\cares.lib address_sorting.lib wsock32.lib ws2_32.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
   Creating library grpc_csharp_ext.lib and object grpc_csharp_ext.exp
crypto.lib(aesni-x86_64.asm.obj) : error LNK2001: unresolved external symbol BORINGSSL_function_hit
```
https://source.cloud.google.com/results/invocations/997a418a-0d57-4957-97ed-a03385f9936d/targets/github%2Fgrpc/tests

**Comment by davidben** [comment]
Once gRPC get past this backlog, we really need to set something up so this happens again. Things are much smoother when you update things regularly. You also get to be part of our feedback loop, so if some change is too disruptive we can come up with alternatives. If you wait over a year, it's far too late to change anything. It also means that, when unexpected breakage happens (like whatever tickled the weird Obj-C issue), it's plausible to identify the trigger.

> /Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/objective-c/tests/Pods/Headers/Private/BoringSSL-GRPC/openssl_grpc/ssl.h:142:2: error: unterminated conditional directive
> #ifndef OPENSSL_HEADER_SSL_H

That ifdef is terminated just fine at the end of the header. We would have entirely caught on fire if that weren't the case. I assume Obj-C is doing something very weird here. I don't know much about that one so you may need to debug it.

> Also found that the python linux artifact only fails for x86 artifacts (the 64bit ones are passing)

That makes sense. SSE2 is mandatory in 64-bit and, per the discussion earlier, the GCC bug kicks in when the compiler *doesn't* get the assume SSE2. You probably can workaround this for now by building with `-msse2` but you really need to get on a newer toolchain.

> Got over the cmake build problem (by removing add_definitions() for msvc warnings) but now running into another problem with Windows C# artifact build:

Right, this is because of some tests which expect the `-DNDEBUG` flag to match between the C files and the assembly files. This was documented in the commit message's [Update-Note marker](https://boringssl.googlesource.com/boringssl/+/c1615719cec110640e2408c78272c0f8db21882f). The tooling for the consumers will look for that marker to help out with changes with a [higher breakage risk](https://boringssl.googlesource.com/boringssl/+/HEAD/BREAKING-CHANGES.md).

For CMake, stick this somewhere:
```
if(CMAKE_BUILD_TYPE STREQUAL "Release")
  # Windows release builds don't set NDEBUG in NASM flags automatically.
  set(CMAKE_ASM_NASM_FLAGS "${CMAKE_ASM_NASM_FLAGS} -DNDEBUG")
endif()
```

**Comment by jtattermusch** [comment]
Current status:
1. it looks like the python linux artifact build now passes (after applying the -msse2 workaround)

2. C# windows artifact is still broken, wit the same error: `crypto.lib(aesni-x86_64.asm.obj) : error LNK2001: unresolved external symbol BORINGSSL_function_hit` error - applying the NDEBUG flag to nasm build doesn't seem to be helping.

3. ObjC build still fails with `ssl.h:142:2: error: unterminated conditional directive
#ifndef OPENSSL_HEADER_SSL_H` - I have no idea what's wrong there.

**Comment by jtattermusch** [comment]
@muxi  can you take a look at what could be wrong with 3.) in my previous comment? I tried, but I'm having a hard time building objC locally.

**Comment by muxi** [comment]
Ok. Looking

File: src/objective-c/BoringSSL-GRPC.podspec [hunk 1]
```diff
@@ -1547,7 +1547,7 @@ Pod::Spec.new do |s|
     EOF
 
     sed -i'.back' '/^#define \\([A-Za-z0-9_]*\\) \\1/d' include/openssl/ssl.h
-    sed -i'.back' 'N;/^#define \\([A-Za-z0-9_]*\\) *\\\\\\n *\\1/d' include/openssl/ssl.h
+    sed -i'.back' '$!N;/^#define \\([A-Za-z0-9_]*\\) *\\\\\\n *\\1/d' include/openssl/ssl.h
```
Comment (jtattermusch) [inline src/objective-c/BoringSSL-GRPC.podspec]: can you explain that a bit? Btw, this part probably deserves a comment because it's pretty cryptic.

**Comment by jtattermusch** [comment]
@muxi  thanks for help. 
Now I'm getting this error:
```
/Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/test/cpp/ios/Pods/BoringSSL-GRPC/err_data.c:22:26: error: unknown type name 'ERR_LIB_NONE'
  OPENSSL_COMPILE_ASSERT(ERR_LIB_NONE == 1, library_values_changed_1);
```
https://source.cloud.google.com/results/invocations/7fbd901f-86b8-46d1-94e0-ecb2d866f242/targets/github%2Fgrpc%2Frun_tests%2Fobjc_macos_opt_native%2Fios-cpp-test-cronet/tests

File: src/objective-c/BoringSSL-GRPC.podspec (hunk 1 above)
Comment (jtattermusch) [reply src/objective-c/BoringSSL-GRPC.podspec]: Ok, I found this one https://github.com/grpc/grpc/blob/6571386a06a714b9c04c8af8c03b063e5efc23a8/templates/src/objective-c/BoringSSL-GRPC.podspec.template#L210
IMHO this is a super dirty hack. Can we come up with a better solution?

Comment (muxi) [reply src/objective-c/BoringSSL-GRPC.podspec]: This was the best way I found under the current strategy. I can make some exploration on using `BORINGSSL_PREFIX_SYMBOLS`, but for now I'm not sure if that will satisfy our need, and it takes cycles. I would try making the current strategy work first.

**Comment by jtattermusch** [comment]
@muxi you fix seems to have helped, but now getting yet another error
```
/Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/objective-c/examples/Sample/Pods/BoringSSL-GRPC/ssl/t1_lib.cc:3434:9: error: use of undeclared identifier 'SSL_get_secure_renegotiation_support'; did you mean 'GRPC_SHADOW_SSL_get_secure_renegotiation_support'?
      !(SSL_get_secure_renegotiation_support(ssl) &&
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        GRPC_SHADOW_SSL_get_secure_renegotiation_support
In file included from /Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/objective-c/examples/Sample/Pods/BoringSSL-GRPC/ssl/t1_lib.cc:109:
/Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/objective-c/examples/Sample/Pods/BoringSSL-GRPC/include/openssl/ssl.h:4617:5: note: expanded from macro 'SSL_get_secure_renegotiation_support'
    SSL_get_secure_renegotiation_support
    ^
/Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/objective-c/examples/Sample/Pods/BoringSSL-GRPC/include/openssl/ssl.h:1565:20: note: 'GRPC_SHADOW_SSL_get_secure_renegotiation_support' declared here
OPENSSL_EXPORT int SSL_get_secure_renegotiation_support(const SSL *ssl);
                   ^
```

**Comment by muxi** [comment]
yep I'm looking at it

**Comment by muxi** [comment]
Looks like the script is ok now. I don't really get the sanity check failure; I do need formatting at that place. Is there any way to suppress formatter for a line or a file?

**Comment by jtattermusch** [comment]
> Looks like the script is ok now. I don't really get the sanity check failure; I do need formatting at that place. Is there any way to suppress formatter for a line or a file?

I fixed that (see my commit), thanks for fixing the objc build (I still think that boringssl podspec hack is horrible though ;-) )

**Comment by muxi** [comment]
That's genius fix :) 
>I still think that boringssl podspec hack is horrible though

I agree with that. I'll check if `BORINGSSL_PREFIX_SYMBOLS` is a valid alternative

**Comment by jtattermusch** [comment]
Ok, looks like all test failures are resolved, except the `error LNK2001: unresolved external symbol BORINGSSL_function_hit` in windows artifacts build.

@davidben  it seems that the workaround I added (defining NDEBUG for NASM) https://github.com/grpc/grpc/pull/20100/commits/90509fcb68d594b38a5c6a10b5d7014b5f25416c is not working for some reason.

**Comment by davidben** [comment]
Hrmf. Alright, I'll see about just making the `BORINGSSL_function_hit` tests opt-in then. Getting `-DNDEBUG` passed into nasm may be more of a headache than is worth it. :-/

**Comment by davidben** [comment]
Waiting on review, but https://boringssl-review.googlesource.com/c/boringssl/+/37764 should do it.

**Comment by jtattermusch** [comment]
@davidben thanks! once the fix has been backported to chromium-stable branch, I can update this PR (hopefully for the last time) and then we should be able to finally merge.
After merging I'll try to add regular boringssl updates to our release process so that we stay up to date and prevent painful updates.

**Comment by davidben** [comment]
Ugh. I think we need to revisit whether it's acceptable for you all to use chromium-stable instead of master, as is standard for our consumers. Given the number of complications around your use of BoringSSL and the vast technical debt we need to clear around it, I do not think the delayed feedback loop is tenable.

We'll can figure something out for the time being, but this change thoroughly does not meet the criteria for a Chromium post-stable merge. (Unlike the last one, it does affect files that Chromium uses.) E.g., as this change will not be merged to Chromium, it's quite likely we'll forget in 6 weeks when chromium-stable moves to another branch.

**Comment by jtattermusch** [comment]
> Ugh. I think we need to revisit whether it's acceptable for you all to use chromium-stable instead of master, as is standard for our consumers. Given the number of complications around your use of BoringSSL and the vast technical debt we need to clear around it, I do not think the delayed feedback loop is tenable.
> 
> We'll can figure something out for the time being, but this change thoroughly does not meet the criteria for a Chromium post-stable merge. (Unlike the last one, it does affect files that Chromium uses.) E.g., as this change will not be merged to Chromium, it's quite likely we'll forget in 6 weeks when chromium-stable moves to another branch.

@davidben  I think it's fine for use to update to the master branch this time (I'll need to test if that works) before we figure out the process for what we should do in the long run.

**Comment by jtattermusch** [comment]
@davidben   is the fix https://boringssl-review.googlesource.com/c/boringssl/+/37764 correct?
From 
https://github.com/google/boringssl/commit/0de64a749b344ce16367bc6514a7d4549bba8c6c#diff-af3b638bc2a3e6c650974192a53c7291R65 it seems that `-DBORINGSSL_DISPATCH_TEST` is used whenever CMAKE_BUILD_TYPE is not "rel", but on windows the name of release build is "Release". Also, we are actually building our artifacts under "RelWithDebInfo" build (which would also get the  -DBORINGSSL_DISPATCH_TEST flag set (which doesn't seem to be intended).

**Comment by davidben** [comment]
MATCHES is a regex that does a substring match, so it should hit both. And since we always pass the value into both, the important thing is it's always consistent.

Is it not working?

**Comment by davidben** [comment]
Also there's a `NOT` there so it won't get set in `RelWithDebInfo`. Are you confusion that with `RelWithAsserts`? Those are not the same. The latter is something we added.

**Comment by davidben** [comment]
Right, our documented policy is that we require the latest stable version of Go for the standalone developer-focused CMake build. In particular, our tests use the newly-added ed25519 support in the Go standard library.

Our usual downstream setups avoid Perl and Go build dependencies altogether. It sounds like this is some more technical debt in the gRPC build we need to clear after this is through.

**Comment by jtattermusch** [comment]
I was able to fix the Windows cmake build problem by upgrading golang to 1.13.1.

Now, I'm getting a problem with with the ObjC build (hopefully this is the very last item to resolve).
```
BoringSSL-GRPC/ssl/ssl_key_share.cc:34:10: fatal error: '../third_party/sike/sike.h' file not found
#include "../third_party/sike/sike.h"
```

**Comment by davidben** [comment]
(For later) I feel like there should be a way to make that podspec file consume the source list in some convenient format or other... :-)

I'm thinking once we get this cleared I'll revise the master-with-bazel branch (or make a new one) that pregenerates all the downstream source lists, not just bazel. And then we add new source list formats as needed for you all, as well as teach the standalone CMake build to run in a mode that consumes the pregenerated files, so it doesn't need Go. And then we need to do something about the symbol prefixing scheme.

I'd like to get to the point that updating BoringSSL for you all is purely a submodule update most of the time. Though it'll probably be a multistep process.

**Comment by jtattermusch** [comment]
> (For later) I feel like there should be a way to make that podspec file consume the source list in some convenient format or other... :-)

Yes, I with there was. Unfortunately our objC build is full of hacks - hopefully we can fix that over time.
> 
> I'm thinking once we get this cleared I'll revise the master-with-bazel branch (or make a new one) that pregenerates all the downstream source lists, not just bazel. 

I think that would be pretty useful.

And then we add new source list formats as needed for you all, as well as teach the standalone CMake build to run in a mode that consumes the pregenerated files, so it doesn't need Go and Perl. 

A standalone CMake build that can consume the pre-generated files would be great - I was going to propose that. It seems that adding on option to just use the pre-generated files shouldn't be that har.d

And then we need to do something about the symbol prefixing scheme.
> 
> I'd like to get to the point that updating BoringSSL for you all is purely a submodule update most of the time. Though it'll probably be a multistep process.

+1 let's sync up after we finally land this PR.

**Comment by muxi** [comment]
@davidben - how does [this line](https://github.com/google/boringssl/blob/1458b49a9e53b0db0ad63e8ef3156214f1473d87/third_party/sike/utils.h#L12) work in BoringSSL? There is not a file `boringssl/third_party/sike/../internal.h`

**Comment by agl** [comment]
> how does this line work in BoringSSL?

Sorry, our mistake. I'm assuming that BoringSSL's top-level `include` path is in the search path usually and the file is found relative to that. [This change](https://boringssl-review.googlesource.com/c/boringssl/+/38344) is pending to fix it.

**Comment by muxi** [comment]
Thanks for the fix Adam. I'll update our reference once your fix is merged.

**Comment by muxi** [comment]
@agl's fix works. Currently the problem is that `sike` contains some perl generated assembly. The assembly is unable to be built by Xcode (clang). Still looking for a solution.

**Comment by agl** [comment]
The `sike` code is probably ready to be removed anyway. Let me do that this afternoon and save worrying about getting it building in Xcode.

**Comment by muxi** [comment]
That would be awesome. Thank you!

**Comment by agl** [comment]
https://boringssl-review.googlesource.com/c/boringssl/+/38384 is running the commit queue now and removes the SIKE code.

**Comment by agl** [comment]
(Now landed.)

**Comment by muxi** [comment]
Seems to be working on my machine. Thank you :)

**Comment by jtattermusch** [comment]
Known issues:
https://github.com/grpc/grpc/issues/20436
https://github.com/grpc/grpc/issues/20694

**Comment by jtattermusch** [comment]
Finally merging, thanks everyone for help. I'll followup on how we can improve the situation in the future.

**Comment by nanahpang** [comment]
@jtattermusch This PR is causing errors:
//depot/google3/third_party/grpc/src/boringssl/err_data.c
1.) The source code file "//depot/google3/third_party/grpc/src/boringssl/err_data.c" contains the license(s) "ISC", which is missing from the LICENSE file.
The LICENSE file "//depot/google3/third_party/grpc/LICENSE" contains the licenses "Apache-2.0".

//depot/google3/third_party/grpc/src/boringssl/crypto_test_data.cc
1.) The source code file "//depot/google3/third_party/grpc/src/boringssl/crypto_test_data.cc" contains the license(s) "ISC", which is missing from the LICENSE file.
The LICENSE file "//depot/google3/third_party/grpc/LICENSE" contains the licenses "Apache-2.0".

**Comment by davidben** [comment]
Those files and licenses aren't new. On my phone so I can't check, but is it possible that they flagged last time BoringSSL was updated and they were ignored? Or perhaps you all manually removed them on import last time and forgot that this is yet another manual step?

Part of the trouble with not updating for a year is things like this tend to get forgotten. :-)

**Comment by nanahpang** [comment]
@davidben Okay, I see. But I couldn't find the process in the import documents. Is there a link you could point me to, or do you know what manual steps are required? Thanks for your help :)

**Comment by nanahpang** [comment]
Also, does the build failure of Objective-C targets related to this PR? I assume it does because no other PR makes changes to the compiler. So I point the failure here. Let me know if I am wrong.

**Comment by jtattermusch** [comment]
@nanahpang  please provide links for those failures, otherwise it's hard to investigate.

**Comment by jtattermusch** [comment]
@nanahpang  let's discuss internally.

**Comment by jtattermusch** [comment]
Looks like this PR also broke some of the older ruby distribtests:
https://source.cloud.google.com/results/invocations/bf193638-559f-45f5-8d5d-6025927fac3c/targets/github%2Fgrpc/tests;group=tasks;test=distribtest.ruby_linux_x64_ubuntu1204_version_unspecified;row=2

**Comment by nanahpang** [comment]
Sure, let's discuss internally. There are some internal links related.

**Pull request merged**